xmtp_mls/src/groups/welcomes/validated_membership.rs

```diff
@@ -54,22 +53,50 @@ where
             .collect();
         let results = futures::future::try_join_all(futures).await?;
 
+        let mut expected_members = HashMap::<String, HashSet<Vec<u8>>>::new();
         for association_state in results {
-            expected_installation_ids.extend(association_state.installation_ids());
+            expected_members.insert(
+                association_state.inbox_id().to_string(),
+                HashSet::from_iter(association_state.installation_ids()),
+            );
         }
 
-        let actual_installation_ids: HashSet<Vec<u8>> = staged_welcome
-            .public_group()
-            .members()
-            .map(|member| member.signature_key)
-            .collect();
-
-        // exclude failed installations
-        expected_installation_ids.retain(|id| !membership.failed_installations.contains(id));
-
-        if expected_installation_ids != actual_installation_ids {
-            return Err(GroupError::InvalidGroupMembership);
+        for member in staged_welcome.public_group().members() {
+            let basic_credential = BasicCredential::try_from(member.credential.clone())?;
+            let claimed_inbox_id = parse_credential(basic_credential.identity())?;
+            if !expected_members.contains_key(&claimed_inbox_id) {
+                tracing::error!(
+                    claimed_inbox_id = claimed_inbox_id,
+                    "Inbox ID not found in expected members",
+                );
+                return Err(GroupError::InvalidGroupMembership);
+            }
+            let installation_ids = expected_members.get_mut(&claimed_inbox_id).unwrap();
+            if !installation_ids.contains(&member.signature_key) {
+                tracing::error!(
+                    claimed_inbox_id = claimed_inbox_id,
+                    installation_id = hex::encode(member.signature_key),
+                    "Installation ID not found in expected members for inbox ID",
+                );
+                return Err(GroupError::InvalidGroupMembership);
+            }
+            installation_ids.remove(&member.signature_key);
+        }
+        for installation_set in expected_members.values() {
+            for remaining_installation_id in installation_set {
+                if !membership
+                    .failed_installations
+                    .contains(remaining_installation_id)
+                {
+                    tracing::error!(
+                        installation_id = hex::encode(remaining_installation_id),
+                        "Installation ID not found in expected members for inbox ID",
+                    );
+                    return Err(GroupError::InvalidGroupMembership);
+                }
+            }
         }
+        // TODO: Is it an error if there are 'failed installations' that are not in the expected members list?
```

Comment:
We don't return an error in this situation currently, so I'm preserving the existing behavior to keep this PR small, but it's worth considering whether or not it's a problem that we don't catch this today.

It's a bit complicated to think through scenarios such as an installation being added to the 'failed installations', and then being revoked from the inbox later on.

Reply:
Sounds good, I will investigate as part of the next PR